src/vos/vos_iterator.c

```diff
@@ -669,63 +662,127 @@
+	if (validate_rc != 0) {
+		/** In theory, this should never happen with our use cases but let's
+		 *  warn if it does.
+		 */
+		D_WARN("Iterator tree disappeared above iterator level %d at %d\n", iter->it_type,
+		       validate_rc);
+	}
+}
+
 static inline int
 vos_iter_cb(vos_iter_cb_t iter_cb, daos_handle_t ih, vos_iter_entry_t *iter_ent,
-	    vos_iter_type_t type, vos_iter_param_t *param, void *arg, unsigned int *acts)
+	    vos_iter_type_t type, vos_iter_param_t *param, void *arg, unsigned int *acts,
+	    struct vos_iter_anchors *anchors)
 {
 	struct vos_iterator *iter = vos_hdl2iter(ih);
-	int		rc;
+	int                  rc;
 
 	vos_iter_sched_sync(iter);
 	D_ASSERT(iter_cb != NULL);
 	rc = iter_cb(ih, iter_ent, type, param, arg, acts);
-	if (vos_iter_sched_check(iter))
+	if (vos_iter_sched_check(iter)) {
 		*acts |= VOS_ITER_CB_YIELD;
+		if (iter->it_parent == NULL || rc != 0 || iter->it_type <= VOS_ITER_OBJ ||
+		    anchors->ia_probe_level != 0)
+			return rc;
+		validate_internal(iter);
+	}
 
 	return rc;
 }
 
-#define JUMP_TO_STAGE(rc, next_label, probe_label, abort_label)				\
-	do {										\
-		switch (rc) {								\
-		case ITER_ABORT:							\
-			(rc) = 0;							\
-			/* fallthrough */						\
-		case ITER_EXIT:								\
-			goto abort_label;						\
-		case ITER_NEXT:								\
-			goto next_label;						\
-		case ITER_PROBE:							\
-			goto probe_label;						\
-		case ITER_CONTINUE:							\
-			(rc) = 0;							\
-			break;								\
-		default:								\
-			D_ASSERTF(rc == 0, "Unexpected positive return code: %d\n",	\
-				  rc);							\
-		}									\
+#define JUMP_TO_STAGE(rc, next_label, probe_label, abort_label)                                    \
+	do {                                                                                       \
+		switch (rc) {                                                                      \
+		case ITER_ABORT:                                                                   \
+			(rc) = 0;                                                                  \
+			/* fallthrough */                                                          \
+		case ITER_EXIT:                                                                    \
+			goto abort_label;                                                          \
+		case ITER_NEXT:                                                                    \
+			goto next_label;                                                           \
+		case ITER_PROBE:                                                                   \
+			goto probe_label;                                                          \
+		case ITER_CONTINUE:                                                                \
+			(rc) = 0;                                                                  \
+			break;                                                                     \
+		default:                                                                           \
```

Comment:
(style)  Possible switch case/default not preceeded by break or fallthrough comment